addons/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/handler/ModbusDataThingHandler.java

```diff
@@ -0,0 +1,798 @@
+        if (bitQuery && readSubIndex.isPresent()) {
+            String errmsg = String.format("readStart=X.Y is not allowed to be used with coils or discrete inputs!");
+            logger.error("Thing '{}' invalid readStart: {}", getThing().getUID(), errmsg);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, errmsg);
+            return false;
+        }
+
+        if (valueTypeBitCount >= 16 && readSubIndex.isPresent()) {
+            String errmsg = String
+                    .format("readStart=X.Y is not allowed to be used with value types larger than 16bit!");
+            logger.error("Thing '{}' invalid readStart: {}", getThing().getUID(), errmsg);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, errmsg);
+            return false;
+        } else if (!bitQuery && valueTypeBitCount < 16 && !readSubIndex.isPresent()) {
+            String errmsg = String.format("readStart=X.Y must be used with value types less than 16bit!");
+            logger.error("Thing '{}' invalid readStart: {}", getThing().getUID(), errmsg);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, errmsg);
+            return false;
+        } else if (readSubIndex.isPresent() && (readSubIndex.get() + 1) * valueTypeBitCount > 16) {
+            // the sub index Y (in X.Y) is above the register limits
+            String errmsg = String.format("readStart=X.Y, the value Y is too large");
+            logger.error("Thing '{}' invalid readStart: {}", getThing().getUID(), errmsg);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, errmsg);
+            return false;
+        }
+
+        // Determine bit positions polled, both start and end inclusive
+        int pollStartBitIndex = pollTask.getRequest().getReference() * dataElementBits;
+        int pollEndBitIndex = pollStartBitIndex + pollTask.getRequest().getDataLength() * dataElementBits;
+
+        // Determine bit positions read, both start and end inclusive
+        int readStartBitIndex = readIndex.get() * dataElementBits + readSubIndex.orElse(0) * valueTypeBitCount;
+        int readEndBitIndex = readStartBitIndex + valueTypeBitCount - 1;
+
+        if (readStartBitIndex < pollStartBitIndex || readEndBitIndex > pollEndBitIndex) {
+            String errmsg = String.format(
+                    "Out-of-bounds: Poller is reading from index %d to %d (inclusive) but this thing configured to read '%s' starting from element %d. Exceeds polled data bounds.",
+                    pollStartBitIndex / dataElementBits, pollEndBitIndex / dataElementBits, readValueType,
+                    readIndex.get());
+            logger.error("Thing {} '{}' readIndex is out of bounds: {}", getThing().getUID(), getThing().getLabel(),
+                    errmsg);
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.CONFIGURATION_ERROR, errmsg);
```

Comment:
It's configuration, I think, since the error can be determined without having any communication with the devices.

It's a traditional out-of-bounds situation when we know size of the array and index at check-time.